File: Go/2_addTwoNumbers/2_addTwoNumbers.go
package q002

import (
	"fmt"
)

type listnode struct {
	val	int
	next *listnode
}

func addTwoNumbers(l1 *listnode, l2 *listnode) *listnode {
	head := &listnode{0, nil}
	curr := head

	p := l1
	q := l2
	carry := 0
	for p != nil || q != nil {
		pVal := 0
		if p != nil {
			pVal = p.val
		}
		qVal := 0
		if q != nil {
			qVal = q.val
		}

		sum := pVal + qVal + carry
		curr.next = &listnode{sum % 10, nil}
		curr = curr.next
		carry = sum / 10

		if p != nil {
			p = p.next
		}
		if q != nil {
			q = q.next
		}
	}

	if carry > 0 {
		curr.next = &listnode{1, nil}
		curr = curr.next
	}

	return head.next
}

// Test ...
func Test() {
	fmt.Println("2_addTwoNumbers begin")
	
	l13 := listnode {3, nil}
	l12 := listnode {4, &l13}
	l11 := listnode {2, &l12}
	l23 := listnode {7, nil}
	l22 := listnode {6, &l23}
	l21 := listnode {5, &l22}

	l := addTwoNumbers(&l11, &l21)
	for iter := l; iter != nil; iter = iter.next {
		fmt.Printf("%d ", iter.val)
	}
	fmt.Println()
	
	fmt.Println("2_addTwoNumbers end")
}

File: CPP/2_addTwoNumbers/2_addTwoNumbers.cpp
// 2_addTwoNumbers.cpp : 定义控制台应用程序的入口点。
//

#include "stdafx.h"
#include <iostream>
using namespace std;

struct ListNode
{
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

ListNode* addTwoNumbers(ListNode* l1, ListNode* l2)
{
	/*ListNode* p = l1;
	ListNode* q = l2;
	ListNode* l = NULL;
	ListNode* r = NULL;

	int carry = 0;
	while (p != NULL && q != NULL)
	{
		ListNode* temp = new ListNode((p->val + q->val + carry) % 10);
		if (r == NULL)
		{
			r = temp;
			l = r;
		}
		else
		{
			r->next = temp;
			r = r->next;
		}

		carry = (p->val + q->val + carry) / 10;
		p = p->next;
		q = q->next;
	}

	while (p != NULL)
	{
		ListNode* temp = new ListNode((p->val + carry) % 10);
		if (r == NULL)
		{
			r = temp;
			l = r;
		}
		else
		{
			r->next = temp;
			r = r->next;
		}

		p = p->next;
	}
		
	while (q != NULL)
	{
		ListNode* temp = new ListNode((q->val + carry) % 10);
		if (r == NULL)
		{
			r = temp;
			l = r;
		}
		else
		{
			r->next = temp;
			r = r->next;
		}

		q = q->next;
	}

	if (carry != 0)
	{
		ListNode* temp = new ListNode(carry % 10);
		if (r == NULL)
		{
			r = temp;
			l = r;
		}
		else
		{
			r->next = temp;
			r = r->next;
		}
	}*/

	ListNode* head = &ListNode(0);
	ListNode* curr = head;
	ListNode* p = l1;
	ListNode* q = l2;
	int carry = 0;
	while (p != NULL || q != NULL)
	{
		int pVal = (p != NULL ? p->val : 0);
		int qVal = (q != NULL ? q->val : 0);
		int sum = pVal + qVal + carry;
		curr->next = new ListNode(sum % 10);
		curr = curr->next;
		carry = sum / 10;

		if (p != NULL) p = p->next;
		if (q != NULL) q = q->next;
	}

	if (carry > 0)
	{
		curr->next = new ListNode(1);
		curr = curr->next;
	}

	return head->next;
}

int _tmain(int argc, _TCHAR* argv[])
{
	ListNode l11 = ListNode(2);
	ListNode l12 = ListNode(4);
	ListNode l13 = ListNode(3);
	l11.next = &l12;
	l12.next = &l13;

	ListNode l21 = ListNode(5);
	ListNode l22 = ListNode(6);
	ListNode l23 = ListNode(7);
	l21.next = &l22;
	l22.next = &l23;

	ListNode* l = addTwoNumbers(&l11, &l21);
	ListNode* p = l;
	ListNode* q = p;
	while (p != NULL)
	{
		cout << p->val << " ";
		
		q = p;
		p = p->next;
		delete q;
	}
	cout << endl;

	system("pause");
	return 0;
}

